### PR TITLE
[DELLEMC][S6000] Platform 2.0 Reboot Reason and fixed issue in process-reboot-cause

### PR DESCRIPTION
**Pull request by sridhar-ravindran**
**- What I did**
1) Added Reboot Reason for S6000 in platform 2.0
2) Fixed issue in process-reboot-cause
3) Added package uninstall code in platform de-init code for z9100, s6100
**- How I did it**
    -> Added support for S6000 Reboot Reason
    -> Added platform.py for all platforms
    -> Verified show reboot-cause command with the code changes. Added UT logs with show reboot-cause
    -> Modified process-reboot-cause service to start after pmon.service. In S6000, we have to wait for nvram to be loaded.
    -> If reboot-cause service starts before pmon.service, show reboot-cause is showing incorrect reason.
    -> Bug fix in process-reboot-cause file
           - import sonic_platform
           + import sonic_platform.platform

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   files/image_config/process-reboot-cause/process-reboot-cause
        modified:   files/image_config/process-reboot-cause/process-reboot-cause.service
        modified:   platform/broadcom/sonic-platform-modules-dell/debian/platform-modules-s6000.init
        modified:   platform/broadcom/sonic-platform-modules-dell/debian/platform-modules-s6000.install
        modified:   platform/broadcom/sonic-platform-modules-dell/debian/rules
        modified:   platform/broadcom/sonic-platform-modules-dell/s6000/modules/dell_s6000_platform.c
        copied:     platform/broadcom/sonic-platform-modules-dell/debian/platform-modules-s6000.init -> platform/broadcom/sonic-platform-modules-dell/s6000/scripts/s6000_platform.sh
        new file:   platform/broadcom/sonic-platform-modules-dell/s6000/setup.py
        new file:   platform/broadcom/sonic-platform-modules-dell/s6000/sonic_platform/__init__.py
        new file:   platform/broadcom/sonic-platform-modules-dell/s6000/sonic_platform/chassis.py
        new file:   platform/broadcom/sonic-platform-modules-dell/s6000/sonic_platform/platform.py
        modified:   platform/broadcom/sonic-platform-modules-dell/s6000/systemd/platform-modules-s6000.service
        modified:   platform/broadcom/sonic-platform-modules-dell/s6100/scripts/s6100_platform.sh
        new file:   platform/broadcom/sonic-platform-modules-dell/s6100/sonic_platform/platform.py
        modified:   platform/broadcom/sonic-platform-modules-dell/z9100/scripts/z9100_platform.sh
        new file:   platform/broadcom/sonic-platform-modules-dell/z9100/sonic_platform/platform.py

**- How to verify it**
  Verified show reboot-cause command with the code changes. Added UT logs with show reboot-cause

**- Description for the changelog**

  Added Reboot reason for S6000 and fix issue in process-reboot-cause

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by sridhar-ravindran** [comment]
[Show_Reboot_Reason_UT_logs.txt](https://github.com/Azure/sonic-buildimage/files/3390968/Show_Reboot_Reason_UT_logs.txt)

File: files/image_config/process-reboot-cause/process-reboot-cause [hunk 1]
```diff
@@ -70,7 +70,7 @@ def main():
     # if there is no sonic_platform package installed, we only provide
     # software-related reboot causes.
     try:
-        import sonic_platform
+        import sonic_platform.platform
```
Comment (jleveque) [inline files/image_config/process-reboot-cause/process-reboot-cause]: Why is this necessary?

File: files/image_config/process-reboot-cause/process-reboot-cause.service [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [Unit]
 Description=Reboot cause determination service
-After=rc-local.service
+After=pmon.service
```
Comment (jleveque) [inline files/image_config/process-reboot-cause/process-reboot-cause.service]: Why change this?

File: files/image_config/process-reboot-cause/process-reboot-cause (hunk 1 above)
Comment (sridhar-ravindran) [reply files/image_config/process-reboot-cause/process-reboot-cause]: Hi Joe,

sonic_platform is directory. platform is the file.

When we import directory alone We get the following error.
**From File :- /usr/bin/process-reboot-cause**
    import sonic_platform
    # Check if the previous reboot was caused by hardware
    platform = sonic_platform.platform.Platform()
    chassis = platform.get_chassis()
    hardware_reboot_cause, optional_details = chassis.get_reboot_cause()

root@sonic:/home/admin# /usr/bin/process-reboot-cause
Traceback (most recent call last):
  File "/usr/bin/process-reboot-cause", line 121, in <module>
    main()
  File "/usr/bin/process-reboot-cause", line 76, in main
    platform = sonic_platform.platform.Platform()
**AttributeError: 'module' object has no attribute 'platform'**


When we import the file "platform" the script is working
    **import sonic_platform.platform**
    # Check if the previous reboot was caused by hardware
    platform = sonic_platform.platform.Platform()
    chassis = platform.get_chassis()
    hardware_reboot_cause, optional_details = chassis.get_reboot_cause()

root@sonic:/home/admin# /usr/bin/process-reboot-cause
root@sonic:/home/admin#

Please let me if my understanding is fine.

Thanks
Sridhar.R

File: files/image_config/process-reboot-cause/process-reboot-cause.service (hunk 2 above)
Comment (sridhar-ravindran) [reply files/image_config/process-reboot-cause/process-reboot-cause.service]: Hi Joe, 
The hardware drivers required to provide reboot reason are installed only inside platform-modules-<device>  service. 

When process-reboot-cause service starts before the pmon service, loading of drivers are not complete and it leads to race condition where the show reboot-cause CLI return incorrect answers.

If process-reboot-cause loads after pmon.service, the platform drivers would have successfully loaded by then and show reboot-cause cli will return correct value

Thanks
Sridhar.R

Comment (jleveque) [reply files/image_config/process-reboot-cause/process-reboot-cause.service]: Hi Sidhar,

SONiC requires that hardware reboot reason and watchdog functionality need to be available in the host OS. The sonic_platform package should be installed in both the host OS and PMon container. Host OS services should not have any dependencies on the containers. We want reboot cause and watchdog to be as reliable as possible. Thanks.

File: files/image_config/process-reboot-cause/process-reboot-cause (hunk 1 above)
Comment (jleveque) [reply files/image_config/process-reboot-cause/process-reboot-cause]: Hmmm. I don't believe this should be necessary. You should be able to import the entire package by specifying the package (directory) name. Let me look into this further.

File: files/image_config/process-reboot-cause/process-reboot-cause.service (hunk 2 above)
Comment (sridhar-ravindran) [reply files/image_config/process-reboot-cause/process-reboot-cause.service]: Hi Joe,
Thanks for the input. I have removed the pmon dependency from process-restart-cause service. 
Thanks
Sridhar.R

File: files/image_config/process-reboot-cause/process-reboot-cause (hunk 1 above)
Comment (jleveque) [reply files/image_config/process-reboot-cause/process-reboot-cause]: It appears for this to work, you need to modify the __init__.py file of your sonic_platform package to either import the files and/or list the module names in `__all__`. See: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

Comment (sridhar-ravindran) [reply files/image_config/process-reboot-cause/process-reboot-cause]: Hi Joe, 
After some literature study, i was able to successfully make it work. Updated __init__.py with 
__all__ = ["platform", "chassis"]
from sonic_platform import *

On adding these 2 lines, "import sonic_platform" works fine.

On adding new files like sfp.py , psu.py, device,py, we have to update the __init__.py file as well.

Thanks for your help and support Joe.

Thanks
Sridhar.R

**Comment by jleveque** [comment]
Retest vs please

**Comment by sridhar-ravindran** [comment]
Hi, 

Could you please port this changes to november branch and 201904 branch

**Comment by jleveque** [comment]
@sridhar-ravindran: will this PR cherry-pick cleanly into those branches?

**Comment by sridhar-ravindran** [comment]
> @sridhar-ravindran: will this PR cherry-pick cleanly into those branches?

Hi @jleveque 
I will check and get back on this
Thanks
Sridhar.R

**Comment by yxieca** [comment]
This PR cannot be cherry-picked into 201811 branch directly. Can you make an PR 201811?
